src/Rings/mpoly-graded.jl

```diff
@@ -1562,6 +1562,397 @@ end
+    G = vcat(G,  extra_gens)
+  end
+  return G
+end
+
+# Fully general, but typically much slower than specialized method below (for positive gradings)
+# Optional kwarg:
+#   extra_gens_flag: if false, inhibits computing grobner basis speculatively in _gens_for_homog_via_sat
+function _homogenization_via_saturation(I::MPolyIdeal{T},  H::homogenizer; extra_gens_flag::Bool = false) where {T <: MPolyRingElem}
+  P = base_ring(I)
+  @req P == H.P  "Homogenizer not for this ring"
+  if is_zero(I)  # special handling for ideal(0)
+    return ideal(zero(H.P_homog)) # return zero ideal in !!P_homog!!
+  end
+  Hgens = H(_gens_for_homog_via_sat(I, extra_gens_flag))
+  R = parent(Hgens[1]) # or H.P_homog
+  if length(Hgens) == 1  # short-cut for principal ideal
+    return ideal(R, Hgens)
+  end
+  DoSatByProduct = false  # true means saturate by product of homogenizing vars;
+                          # false means saturate successively by each homogenizing var (probably faster, in most cases?)
+  HVars = homogenizing_variables(H)
+  if DoSatByProduct
+    Ih = saturation(ideal(R, Hgens), ideal(R, prod(HVars)))
+  else # not DoSatByProduct, so do a cascade of saturations
+    Ih = ideal(R, Hgens)
+    for hvar in HVars
+      Ih = saturation(Ih, ideal(R, hvar))
+      # IMPROVE???  if length(gens(Ih)) == 1  return Ih;  end
+    end
+  end
+  return Ih
+end
+
+
+# ============================================
+# 2023-06-30 START: New impl of homogenization
+# By John Abbott, based on K+R Book "Computational Commutative Algebra" (see "justification" below)
+# This code delegates to _homogenization_via_saturation for non-positive gradings
+# ============================================
+
+# _sat_poly_by_var and kronecker_delta are internal auxiliary functions
+# This homogenization impl seems to be usefully faster with positive ZZ^m-grading for m > 1
+
+# Saturate a polynomial by a variable -- equiv to  f/gcd(f,h^deg(f))
+function _sat_poly_by_var(f,h)
+  # ASSUMES h is a variable i.e. monic poly of deg 1 with just 1 term
+  # Is there a better way to implement this?
+  Ih = ideal([h]);
+  while ideal_membership(f, Ih)
+    f = div(f,h); ##f /= h;  but this shorter form does not work (why?)
+  end
+  return f;
+  # This loop is slower (presumably because it computes the remainder always)
+  # while true
+  #     q,r = divrem(f,h);
+  #     if !is_zero(r)
+  #         return q;
+  #     end
+  #     f = q;
+  # end
+end
+
+# This function should be in a file of utilities (JAA thinks)
+function kronecker_delta(i::Int, j::Int)
+  return (i == j) ? 1 : 0
+end
+
+# MAYBE is_positively_graded makes this fn obsolete -- !!NOT SURE!!
+# # Check that W defines a positive grading: namely
+# #     each col contains a non-zero entry, and
+# #     first non-zero going down the col is positive.
+# # Follows defn 4.2.4 from R+K vol 2.
+# function is_positive_grading_matrix(W::Union{ZZMatrix, Matrix{<:IntegerUnion}})
+#   nrows = size(W,1)
```

Comment:
Is this code left here for a purpose? Do we have another copy of it already in another file? 

If we want to delete it, please do so. Otherwise, it would be good to have a hint why this is kept here so that at a later point one can decide more easily whether this can finally go.

Reply:
There are two definitions of positively graded in the literature.  I must verify that using `is_positively_graded` rather than the commented out code does not adversely affect correctness.  I left the code there since some other urgently needed to use the new UI.

Reply:
The function `is_positively_graded` is not sufficient for my case, so I must use my own "more restrictive" code.